lib/model/receiveonlyfolder.go

```diff
@@ -0,0 +1,158 @@
+// Copyright (C) 2014 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package model
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/events"
+	"github.com/syncthing/syncthing/lib/fs"
+	"github.com/syncthing/syncthing/lib/protocol"
+	"github.com/syncthing/syncthing/lib/sync"
+	"github.com/syncthing/syncthing/lib/versioner"
+)
+
+type receiveOnlyFolder struct {
+	// WO folders are really just RW folders where we reject local changes...
+	sendReceiveFolder
+}
+
+func init() {
+	folderFactories[config.FolderTypeReceiveOnly] = newReceiveOnlyFolder
+}
+
+func newReceiveOnlyFolder(model *Model, cfg config.FolderConfiguration, ver versioner.Versioner, mtimeFS *fs.MtimeFS) service {
+	f := &receiveOnlyFolder{
+		sendReceiveFolder{
+			folder: folder{
+				stateTracker: newStateTracker(cfg.ID),
+				scan:         newFolderScanner(cfg),
+				stop:         make(chan struct{}),
+				model:        model,
+			},
+			FolderConfiguration: cfg,
+
+			mtimeFS:   mtimeFS,
+			dir:       cfg.Path(),
+			versioner: ver,
+
+			queue:       newJobQueue(),
+			pullTimer:   time.NewTimer(time.Second),
+			remoteIndex: make(chan struct{}, 1), // This needs to be 1-buffered so that we queue a notification if we're busy doing a pull when it comes.
+
+			errorsMut: sync.NewMutex(),
+
+			initialScanCompleted: make(chan struct{}),
+		},
+	}
+
+	f.configureCopiersAndPullers()
+
+	return f
+}
+
+func (f *receiveOnlyFolder) String() string {
+	return fmt.Sprintf("receiveOnlyFolder/%s@%p", f.folderID, f)
+}
+
+// rejectLocalChanges reverts all local changes
+func (f *receiveOnlyFolder) rejectLocalChanges(fs []protocol.FileInfo) []protocol.FileInfo {
```

Comment:
Yeah sorry, but this just doesn't sound like the right approach to me. I got a big issue with code duplication. This not only doubles the maintenance going forward, but also adds the risk of forgetting to do things in two places and thus causing even more problems.

Is the if statement what bothers you? I can change it up so we have a function in folders that always gets called. Something like "validateLocalChanges". We then would simply overwrite it for receiveonly to do some real action. Or we could even simply call rejectLocalChanges, which won't do anything on the other folder types.